core/server/master/src/main/java/alluxio/master/throttle/DefaultThrottleMaster.java

```diff
@@ -0,0 +1,158 @@
+          MetricsMaster.class);
+
+  /** The Alluxio master process. */
+  private MasterProcess mMasterProcess;
+
+  private ThrottleExecutor mThrottleExecutor;
+
+  /**
+   * The service that performs license check.
+   */
+  @SuppressFBWarnings("URF_UNREAD_FIELD")
+  private Future<?> mThrottleService;
+
+  /**
+   * Creates a new instance of {@link DefaultThrottleMaster}.
+   *
+   * @param registry the master registry
+   * @param masterContext the context for Alluxio master
+   */
+  public DefaultThrottleMaster(MasterRegistry registry, MasterContext masterContext) {
+    super(masterContext, new SystemClock(), ExecutorServiceFactories
+        .cachedThreadPool(Constants.THROTTLE_MASTER_NAME));
+    registry.add(DefaultThrottleMaster.class, this);
+  }
+
+  /**
+   * Sets the master to be used in {@link ThrottleExecutor} for throttling.
+   * This should be called before calling {@link #start(Boolean)}.
+   *
+   * @param masterProcess the Alluxio master process
+   */
+  public void setMaster(MasterProcess masterProcess) {
+    mMasterProcess = masterProcess;
+    mThrottleExecutor = new ThrottleExecutor(mMasterProcess);
+  }
+
+  @Override
+  public Set<Class<? extends Server>> getDependencies() {
+    return DEPS;
+  }
+
+  @Override
+  public String getName() {
+    return Constants.THROTTLE_MASTER_NAME;
+  }
+
+  @Override
+  public void start(Boolean isLeader) throws IOException {
+    super.start(isLeader);
+    Preconditions.checkNotNull(mMasterProcess, "Alluxio master process is not specified");
+    Preconditions.checkNotNull(mThrottleExecutor, "ThrottleExecutor is not specified");
+    if (!isLeader) {
+      return;
+    }
+    LOG.info("Starting {}", getName());
+    mThrottleService = getExecutorService().submit(
+        new HeartbeatThread(HeartbeatContext.MASTER_THROTTLE, mThrottleExecutor,
+            Configuration.getMs(PropertyKey.MASTER_THROTTLE_HEARTBEAT_INTERVAL),
+            Configuration.global(),
+            mMasterContext.getUserState()));
+    LOG.info("{} is started", getName());
+  }
+
+  @Override
+  public Map<ServiceType, GrpcService> getServices() {
+    return new HashMap<>();
+  }
+
+  /**
+   * Collects and saves call home information during the heartbeat.
+   */
+  @ThreadSafe
+  public static final class ThrottleExecutor implements HeartbeatExecutor {
+    private static final Logger LOG = LoggerFactory.getLogger(ThrottleExecutor.class);
+
+    private MasterProcess mMasterProcess;
+    private SystemMonitor mSystemMonitor;
+
+    /**
+     * Creates a new instance of {@link ThrottleExecutor}.
+     *
+     * @param masterProcess the Alluxio master process
+     */
+    public ThrottleExecutor(MasterProcess masterProcess) {
+      mMasterProcess = masterProcess;
+      mSystemMonitor = new SystemMonitor(mMasterProcess);
+    }
+
+    @Override
+    public void heartbeat() throws InterruptedException {
+      if (!mMasterProcess.isGrpcServing()) {
```

Comment:
It might be worth it to track usage even if not the primary master as the others might also encounter performance issues.

Reply:
agree